src/execution_plan/execution_plan.c

```diff
@@ -904,49 +873,31 @@ ExecutionPlan *NewExecutionPlan(ResultSet *result_set) {
 	uint segment_count = array_len(segment_indices);
 	ExecutionPlan **segments = rm_malloc(segment_count * sizeof(ExecutionPlan *));
 	start_offset = 0;
+	OpBase *prev_scope_end = NULL;
+	OpBase *prev_scope_root = NULL;
 	for(int i = 0; i < segment_count; i++) {
 		uint end_offset = segment_indices[i];
 		// Slice the AST to only include the clauses in the current segment.
 		AST *ast_segment = AST_NewSegment(ast, start_offset, end_offset);
 
 		// Construct a new ExecutionPlanSegment.
-		ExecutionPlan *segment = _NewEmptyExecutionPlan();
-		_PopulateExecutionPlan(segment, result_set);
+		ExecutionPlan *segment = ExecutionPlan_NewEmptyExecutionPlan();
+		ExecutionPlan_PopulateExecutionPlan(segment, result_set, prev_scope_root, prev_scope_end);
 		AST_Free(ast_segment); // Free the AST segment.
 
 		segments[i] = segment;
 		start_offset = end_offset;
+		prev_scope_end = prev_scope_root;
+		prev_scope_root = segment->root;
 	}
 
 	QueryCtx_SetAST(ast); // AST segments have been freed, set master AST in QueryCtx.
 
-	// Place filter ops required by first ExecutionPlan segment.
-	if(segments[0]->filter_tree) _ExecutionPlan_PlaceFilterOps(segments[0], NULL);
-
-	OpBase *connecting_op = NULL;
-	OpBase *prev_scope_end = NULL;
-	// Merge segments.
-	for(int i = 1; i < segment_count; i++) {
-		ExecutionPlan *prev_segment = segments[i - 1];
-		ExecutionPlan *current_segment = segments[i];
-
-		OpBase *prev_root = prev_segment->root;
-		connecting_op = ExecutionPlan_LocateFirstOp(current_segment->root,
-													OPType_PROJECT | OPType_AGGREGATE);
-		assert(connecting_op->childCount == 0);
-
-		ExecutionPlan_AddOp(connecting_op, prev_root);
-
-		// Place filter ops required by current segment.
-		if(current_segment->filter_tree) _ExecutionPlan_PlaceFilterOps(current_segment, prev_scope_end);
-
-		prev_scope_end = prev_root; // Track the previous scope's end so filter placement doesn't overreach.
-	}
```

Comment:
Why did you have to replace this logic? I'm guessing an issue with filter placement, but I had expected all issues to be solvable with the approach used at https://github.com/RedisGraph/RedisGraph/blob/master/src/execution_plan/execution_plan.c#L577-L580 .